custom-elements/custom-elements-registry/define.html

```diff
@@ -0,0 +1,230 @@
+<!DOCTYPE html>
+<title>Custom Elements: Element definition</title>
+<link rel="help" href="https://html.spec.whatwg.org/multipage/scripting.html#element-definition">
+<meta name="author" title="Koji Ishii" href="mailto:kojiishi@gmail.com">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<body>
+<div id="log"></div>
+<script>
+'use strict';
+(function () {
+  var testable = false;
+  test(function () {
+    assert_true('customElements' in window, '"window.customElements" exists');
+    assert_true('define' in window.customElements, '"window.customElements.define" exists');
+    testable = true;
+  }, '"window.customElements.define" exists');
+  if (!testable)
+    return;
+
+  // 1. If IsConstructor(constructor) is false,
+  // then throw a TypeError and abort these steps.
+  test(function () {
+    assert_throws(new TypeError, function () {
+      window.customElements.define();
+    });
+  }, 'If no arguments, throw a TypeError');
+  test(function () {
+    assert_throws(new TypeError, function () {
+      window.customElements.define("test-define-one-arg");
+    });
+  }, 'If one argument, throw a TypeError');
+  [
+    [ "undefined", undefined ],
+    [ "null", null ],
+    [ "object", {} ],
+    [ "string", "string" ],
+    [ "arrow function", () => {} ], // IsConstructor returns false for arrow functions
+    [ "method", ({ m() { } }).m ], // IsConstructor returns false for methods
+  ].forEach(function (t) {
+    test(function () {
+      assert_throws(new TypeError, function () {
+        window.customElements.define("test-define-ctor-" + t[0], t[1]);
+      });
+    }, 'If constructor is ' + t[0] + ', throw a TypeError');
+  });
+
+  // 2. If name is not a valid custom element name,
+  // then throw a SyntaxError and abort these steps.
+  var validCustomElementNames = [
+    // [a-z] (PCENChar)* '-' (PCENChar)*
+    // https://html.spec.whatwg.org/multipage/scripting.html#valid-custom-element-name
+    "a-",
+    "a-a",
+    "aa-",
+    "aa-a",
+    "a-.-_",
+    "a-0123456789",
+    "a-\u6F22\u5B57", // Two CJK Unified Ideographs
+    "a-\uD840\uDC0B", // Surrogate pair U+2000B
+  ];
+  var invalidCustomElementNames = [
+    undefined,
+    null,
+    "",
+    "a",
+    "A",
+    "A-",
+    "0-",
+    "a-A",
+    "a-Z",
+    "A-a",
+    // name must not be any of the hyphen-containing element names.
+    "annotation-xml",
+    "color-profile",
+    "font-face",
+    "font-face-src",
+    "font-face-uri",
+    "font-face-format",
+    "font-face-name",
+    "missing-glyph",
+  ];
+  validCustomElementNames.forEach(function (name) {
+    test(function () {
+      window.customElements.define(name, class {});
```

Comment:
I found writing tests for Custom Elements last time that maintainability becomes a problem if you register everything on the window. WDYT about minting iframes and using their windows?